src/freenet/client/filter/TheoraPacketFilter.java

```diff
@@ -1,148 +1,287 @@
 	}
 
-	private long decode32bitIntegerFrom8BitChunks(DataInputStream input) throws IOException {
-		int LEN0 = input.readUnsignedByte();
-		int LEN1 = input.readUnsignedByte();
-		int LEN2 = input.readUnsignedByte();
-		int LEN3 = input.readUnsignedByte();
-		return LEN0|(LEN1 << 8)|(LEN2 << 16)|(LEN3 << 24);
+	private void parseIdentificationHeader(BitInputStream input) throws IOException {
+		Logger.minor(this, "IDENTIFICATION_HEADER");
+
+		// The header packets begin with the header type and the magic number. Validate both.
+		byte[] magicHeader = new byte[1 + magicNumber.length];
+		input.readFully(magicHeader);
+		checkMagicHeader(magicHeader, (byte) 0x80); // -128
+
+		int VMAJ = input.readInt(8);
+		if (VMAJ != 3) throw new UnknownContentTypeException("Header VMAJ: " + VMAJ);
+
+		int VMIN = input.readInt(8);
+		if (VMIN != 2) throw new UnknownContentTypeException("Header VMIN: " + VMIN);
+
+		int VREV = input.readInt(8);
+		if (VREV > 1) throw new UnknownContentTypeException("Header VREV: " + VREV);
+
+		int FMBW = input.readInt(16);
+		if (FMBW == 0) throw new UnknownContentTypeException("Header FMBW: " + FMBW);
+
+		int FMBH = input.readInt(16);
+		if (FMBH == 0) throw new UnknownContentTypeException("Header FMBH: " + FMBH);
+
+		int PICW = input.readInt(24);
+		if (PICW > FMBW * 16) throw new UnknownContentTypeException("Header PICW: " + PICW + "; FMBW: " + FMBW);
+
+		int PICH = input.readInt(24);
+		if (PICH > FMBH * 16) throw new UnknownContentTypeException("Header PICH: " + PICH + "; FMBH: " + FMBH);
+
+		int PICX = input.readInt(8);
+		if (PICX > FMBW * 16 - PICX) throw new UnknownContentTypeException("Header PICX: " + PICX + "; FMBW: " + FMBW + "; PICX: " + PICX);
+
+		int PICY = input.readInt(8);
+		if (PICY > FMBH * 16 - PICY) throw new UnknownContentTypeException("Header PICY: " + PICY + "; FMBH: " + FMBH + "; PICY: " + PICY);
+
+		int FRN = input.readInt(32);
+		if (FRN == 0) throw new UnknownContentTypeException("Header FRN: " + FRN);
+
+		int FRD = input.readInt(32);
+		if (FRD == 0) throw new UnknownContentTypeException("Header FRN: " + FRN);
+
+		input.skip(48); // skip PARN and PARD
+
+		int CS = input.readInt(8);
+		if (!(CS == 0 || CS == 1 || CS == 2)) throw new UnknownContentTypeException("Header CS: " + CS);
+
+		input.skip(35); // skip NOMBR, QUAL and KFGSHIFT
+
+		int PF = input.readInt(2);
+		if (PF == 1) throw new UnknownContentTypeException("Header PF: " + PF);
+
+		int Res = input.readInt(3);
+		if (Res != 0) throw new UnknownContentTypeException("Header Res: " + Res);
+
+		expectedPacket = Packet.COMMENT_HEADER;
+	}
+
+	private CodecPacket parseCommentHeader(BitInputStream input) throws IOException {
+		Logger.minor(this, "COMMENT_HEADER");
+
+		byte[] magicHeader = new byte[1 + magicNumber.length];
+		input.readFully(magicHeader);
+		checkMagicHeader(magicHeader, (byte) 0x81); // -127
+
+		int vendorLength = input.readInt(32, ByteOrder.LITTLE_ENDIAN);
+		byte[] vendor = new byte[vendorLength];
```

Comment:
And this one I’d also like to check, along with all other occurences of using a user-supplied number to allocate memory. :)

Reply:
Like this allocation not rarely occurs before calling `readFully` method. I launched the search and there are places where the `Integer.MAX_VALUE` can be present according the syntax. But also every pace should be logically checked, is it true.

So, probably this can be registered as an issue in the tracker.

Reply:
Sure, we use the same pattern elsewhere... but in this instance the code is in a content filter: we are parsing presumed-malicious streams... and allocating too much memory doing so will DoS the node (OutOfMemory exceptions...).

Surely the vendor length (and other similar allocations in any of our filters) can be restricted to something sane size-wise ;)

Reply:
At this point we don't know the complete length of the stream being filtered?

Reply:
I'm not confident we can place much of a bound on this, without adding it as an arbitrary limitation. From [the spec](https://www.theora.org/doc/libtheora-1.0alpha6/structtheora__comment.html):

> In filling in this structure, theora_decode_header() will null-terminate the user_comment strings for safety. However, the bitstream format itself treats them as 8-bit clean, and so the length array should be treated as authoritative for their length.

Reply:
A content filter doesn't have to be spec abiding ... it has to be safe.

Reply:
I’m totally with nextgens here. We don’t have to restrict it to 20 chars but something sensible, like, I don’t know, 1k, or maybe 10k shouldn’t catch too many videos (one would have to check with stuff that’s in the wild) and still prevent the file from DoSing the node.

Reply:
So, in this PR I totally removed this place (in COMMENT_HEADER state we check only magicHeader and do not read comments). Additionally I created an issue to check other places https://freenet.mantishub.io/view.php?id=7108. So I am marking this place as resolved at this PR.